Update Limbus

cocosolos opened the pull request:
- Adjust entry requirements.
- Overhaul limbus.lua.
- Update Limbus NPCs.
- Rewrite armoury crate handling.
- Table Limbus mob/npc IDs.
- Move soap timer from Sagheera.
- Extend pathTo to accept pathflags.
- Add special setup for Limbus mobs.

Note that none of the actual BCNM or mob files are adjusted here, and all zones have been locked to prevent entry.

<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

TeoTwawki commented on sql/npc_list.sql:
```diff
@@ -2865,175 +2865,164 @@ INSERT INTO `npc_list` VALUES (16924814,'Mystic_Retriever','Mystic Retriever',13
+INSERT INTO `npc_list` VALUES (16929194,'Matter_Diffusion_Module','Matter Diffusion Module',0,-20.000,-2.500,-544.000,1,50,50,0,0,0,2,3,0x0000340000000000000000000000000000000000,0,'COP',0);
+INSERT INTO `npc_list` VALUES (16929195,'Matter_Diffusion_Module','Matter Diffusion Module',0,300.000,-162.500,-504.000,1,50,50,0,0,0,2,3,0x0000340000000000000000000000000000000000,0,'COP',0);
+INSERT INTO `npc_list` VALUES (16929196,'Matter_Diffusion_Module','Matter Diffusion Module',0,540.000,-2.500,-544.000,1,50,50,0,0,0,2,3,0x0000340000000000000000000000000000000000,0,'COP',0);
+INSERT INTO `npc_list` VALUES (16929197,'_110','',0,340.000,72.000,400.000,1,50,50,8,0,0,0,6147,0x0200000000000000000000000000000000000000,0,'COP',0);
```
could use the empty unused string field to identify it by, if the special name is really required.

cocosolos replied:
No need to click or identify these, just using the IDs to set animation for doors.

ibm2431 commented on scripts/zones/Apollyon/npcs/_127.lua:
```diff
@@ -0,0 +1,41 @@
+
+function onTrade(player, npc, trade)
+    if player:hasKeyItem(tpz.ki.COSMOCLEANSE) and player:hasKeyItem(tpz.ki.RED_CARD) then
+        player:setCharVar("ApollyonEntrance", 0)
```
charVars default to zero if they're not set; I'm still diving into this, but if you're using this later to determine where players should end up, it may present undesired behavior (because later `getCharVar` calls will return 0 when they fail to find a var)

edit: Where is this getting checked? I can't find it 😅

cocosolos replied:
This is checked in the zone.lua to teleport them to the correct side when exiting the battlefield.

cocosolos committed:
Update Limbus

- Adjust entry requirements.
- Overhaul limbus.lua.
- Update Limbus NPCs.
- Rewrite armoury crate handling.
- Table Limbus mob/npc IDs.
- Move soap timer from Sagheera.
- Extend pathTo to accept pathflags.
- Add special setup for Limbus mobs.

ibm2431 commented on src/map/utils/mobutils.cpp:
```diff
@@ -809,9 +809,18 @@ void SetupBattlefieldMob(CMobEntity* PMob)
-    PMob->m_maxRoamDistance = 0.5f;
+    // Limbus mobs
+    uint16 zoneID = PMob->getZone();
+    if(zoneID == 37 || zoneID == 37) 
```
Doesn't have to be done now, but we may want to consider a new zoneType for such zones in the future. I think there's a handful more than just Limbus.

cocosolos replied:
Did this originally and you're probably right. I think this function needs some work anyway (should _any_ bcnm mobs be giving direct exp from kills, for example?)

ibm2431 commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -1798,14 +1798,23 @@ inline int32 CLuaBaseEntity::pathTo(lua_State* L)
 
     position_t point;
-
+    uint8 flags;
```
Maybe move the definition into the check? Save us a whopping two lines. 😉

ibm2431 commented on scripts/zones/Temenos/npcs/Armoury_Crate.lua:
```diff
@@ -815,241 +1179,279 @@ local loot =
+        {
+            [1298] = function() -- Temenos West Crate Handling
+                if crateID ~= ID.npc.TEMENOS_W_CRATE[7] then
+                    for i = 1, 6 do
+                        for j = 0, 2 do
+                            if crateID == ID.npc.TEMENOS_W_CRATE[i]+j then
+                                if model == 960 then
+                                    tpz.battlefield.HealPlayers(battlefield)
+                                elseif model == 961 then
+                                    tpz.limbus.handleLootRolls(battlefield, loot[bfid][i], nil, npc)
+                                elseif model == 962 then
+                                    tpz.limbus.extendTimeLimit(battlefield, 15, tpz.zone.TEMENOS)
+                                end
```
You're working too hard here~

```lua
-- in limbus.lua
tpz.limbus.prepareCrates = function(ids)
    local crates = {}
    for floor, id in ipairs(ids)
        crates[id] = floor
    end
    return crates
end
tpz.limbus.crate[TEMENOS_W] = tpz.limbus.prepareCrates(ID.npc.TEMENOS_W_CRATE)

-- armory_crate.lua
crate_type = model - 960
local baseID = crateID - crate_type
local floor = tpz.limbus.crate[TEMENOS_W][baseID]
if floor then
    if crate_type == 0 then
        tpz.battlefield.HealPlayers(battlefield)
    elseif crate_type == 1 then
        tpz.limbus.handleLootRolls(battlefield, loot[bfid][floor], nil, npc)
        if floor == 7 then
            battlefield:setLocalVar("cutsceneTimer", 10)
            battlefield:setLocalVar("lootSeen", 1)
        end
    elseif crate_type == 2 then
        tpz.limbus.extendTimeLimit(battlefield, 15, tpz.zone.TEMENOS)
    end
end
```

Puts the looping on first require, instead of every time a casket is opened!

cocosolos replied:
This is a weird one, and it may not matter, but all 3 of these zones have different crate mechanics. East has 4 chests and if you get a mimic then there's weird mechanics there, the others have 3. West doesn't depop the other chests, and North has 3 chests and you can only pick 1.

ibm2431 commented on scripts/zones/Temenos/npcs/Armoury_Crate.lua:
```diff
@@ -815,241 +1179,279 @@ local loot =
-            end
+            end,
+            [1299] = function() -- Temenos North Crate Handling
+                if crateID ~= ID.npc.TEMENOS_N_CRATE[7] then
+                    for i = 1, 6 do
+                        for j = 0, 2 do
+                            if crateID == ID.npc.TEMENOS_N_CRATE[i]+j then
+                                if j ~= 0 then GetNPCByID(ID.npc.TEMENOS_N_CRATE[i]):setStatus(tpz.status.DISAPPEAR) end
+                                if j ~= 1 then GetNPCByID(ID.npc.TEMENOS_N_CRATE[i]+1):setStatus(tpz.status.DISAPPEAR) end
+                                if j ~= 2 then GetNPCByID(ID.npc.TEMENOS_N_CRATE[i]+2):setStatus(tpz.status.DISAPPEAR) end
+                                if model == 960 then
+                                    tpz.battlefield.HealPlayers(battlefield)
+                                elseif model == 961 then
+                                    tpz.limbus.handleLootRolls(battlefield, loot[bfid][i], nil, npc)
+                                elseif model == 962 then
+                                    tpz.limbus.extendTimeLimit(battlefield, 15, tpz.zone.TEMENOS)
```
Same logic as above, but just include disappearing the other two crates on the floor after determining which one you're opening.

ibm2431 commented on scripts/zones/Temenos/npcs/Armoury_Crate.lua:
```diff
@@ -815,241 +1179,279 @@ local loot =
+                    for i = 1, 6 do
+                        local mask = battlefield:getLocalVar("crateMaskF"..i)
+                        for j = 0, 3 do
+                            if crateID == ID.npc.TEMENOS_E_CRATE[i]+j then
+                                if GetMobByID(ID.mob.TEMENOS_E_MOB[i]+4):isDead() then
+                                    battlefield:setLocalVar("crateOpenedF"..i, 1)
+                                    if model ~= 961 or not (mask > 7 and spawnMimic) then
+                                        if j ~= 0 then GetNPCByID(ID.npc.TEMENOS_E_CRATE[i]):setStatus(tpz.status.DISAPPEAR) end
+                                        if j ~= 1 then GetNPCByID(ID.npc.TEMENOS_E_CRATE[i]+1):setStatus(tpz.status.DISAPPEAR) end
+                                        if j ~= 2 then GetNPCByID(ID.npc.TEMENOS_E_CRATE[i]+2):setStatus(tpz.status.DISAPPEAR) end
+                                        if j ~= 3 then GetNPCByID(ID.npc.TEMENOS_E_CRATE[i]+3):setStatus(tpz.status.DISAPPEAR) end
+                                    end
+                                    if GetMobByID(ID.mob.TEMENOS_E_MOB[i]):isAlive() then DespawnMob(ID.mob.TEMENOS_E_MOB[i]) end
+                                    if GetMobByID(ID.mob.TEMENOS_E_MOB[i]+1):isAlive() then DespawnMob(ID.mob.TEMENOS_E_MOB[i]+1) end
+                                    if GetMobByID(ID.mob.TEMENOS_E_MOB[i]+2):isAlive() then DespawnMob(ID.mob.TEMENOS_E_MOB[i]+2) end
+                                    if GetMobByID(ID.mob.TEMENOS_E_MOB[i]+3):isAlive() then DespawnMob(ID.mob.TEMENOS_E_MOB[i]+3) end
```
Me: "Hrm, fake chests - this PR is starting to get a little hard to review."
Me: "...Wait, how are we even handling models? They're hard set based on the NPC ID... right?"
_checks_
Me: "Okay, we can do this. We can have random chest models, mimics, pretty code, and without looping. Just... think... and get that review in."
_starts going through Limbus global_
Me: "Nope. Can't get this thoroughly reviewed by Thursday. Punting this to a branch, and I'll open my own PR for logic simplifications. Not gonna make cocosolos put up with my usual nonsense during review of this PR!"

cocosolos replied:
I had this stuff in loops originally, but it was too slow and allowed opening multiple chests by timing correctly or spamming. Some floors change the chest model dynamically.

ibm2431 commented on scripts/globals/bcnm.lua:
```diff
@@ -898,7 +902,23 @@ function EventUpdateBCNM(player, csid, option, extras, entrance)
-
+        switch (battlefieldId): caseof 
+        {
+            [1290] = function() area = 2 end, -- NW_Apollyon
+            [1291] = function() area = 1 end, -- SW_Apollyon
+            [1292] = function() area = 4 end, -- NE_Apollyon
+            [1293] = function() area = 3 end, -- SE_Apollyon
+            [1294] = function() area = 6 end, -- CS_Apollyon
+            [1296] = function() area = 5 end, -- Central_Apollyon
```
Limbus is weird.

I'll want to revisit this ID and area system later.

ibm2431 commented on scripts/zones/Port_Jeuno/npcs/Sagheera.lua:
```diff
@@ -389,7 +389,6 @@ function onEventFinish(player,csid,option)
         local cosmoTime = getCosmoCleanseTime(player)
         if cosmoTime == COSMO_READY and player:delGil(15000) then
             npcUtil.giveKeyItem(player, tpz.ki.COSMOCLEANSE)
-            player:setCharVar("Cosmo_Cleanse_TIME", os.time())
```
Where does this get set? I'm not seeing it in `limbus.enter` or the entry NPCs.

cocosolos replied:
It gets set when you actually enter the bcnm, so next PR.

edit: limbus.enter is just used to teleport you to the correct side of Apollyon.

ibm2431 replied:
Okay, that works!

ibm2431 reviewed (commented):
Thanks for taking this on, cocosolos! 🙇 

I'm seeing the potential for improvements being made to crates so they can be cleaner (generation, opening), but I'll bang out the code myself in a later PR. Not gonna make you worry about that noise.

Usually I make sure every line is reviewed, but I think it'll be more time efficient for me to get hands on with this updated Limbus global and the crate files while trying to "make it prettier" rather than scrolling all over the place on GitHub and typing up code comments~

**Files I viewed and saw no potential problems with:**
 scripts/globals/battlefield.lua
 scripts/zones/AlTaieu/npcs/Swirling_Vortex.lua
 scripts/zones/Apollyon/IDs.lua
 scripts/zones/Apollyon/Zone.lua
 scripts/zones/Apollyon/npcs/Sentinel_Column.lua
 scripts/zones/LaLoff_Amphitheater/IDs.lua
 scripts/zones/LaLoff_Amphitheater/npcs/qm1_1.lua
 scripts/zones/LaLoff_Amphitheater/npcs/qm1_2.lua
 scripts/zones/LaLoff_Amphitheater/npcs/qm1_3.lua
 scripts/zones/LaLoff_Amphitheater/npcs/qm1_4.lua
 scripts/zones/LaLoff_Amphitheater/npcs/qm1_5.lua
 scripts/zones/Temenos/IDs.lua
 scripts/zones/Temenos/Zone.lua
 scripts/zones/Temenos/npcs/Matter_Diffusion_Module.lua
 scripts/zones/Temenos/npcs/Particle_Gate.lua
 scripts/zones/Temenos/npcs/Scanning_Device.lua

**Files I intend to work on later:**
 scripts/globals/limbus.lua
 scripts/zones/Apollyon/npcs/Armoury_Crate.lua
 scripts/zones/Temenos/npcs/Armoury_Crate.lua

Comments/questions on other files below.

cocosolos committed:
Remove exp from BCNM mobs

ibm2431 reviewed (approved):
I'll get hands on with the Limbus global and Armory Crates when this is in a branch~

cocosolos committed:
Don't superlink Limbus mobs
